File: app/src/main/java/views/Graph.java
package views;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.text.TextPaint;
import android.view.View;

import androidx.core.content.ContextCompat;
import androidx.core.content.res.ResourcesCompat;

import com.example.graphmaker.R;

import java.util.ArrayList;

import model.Data;
import model.Player;
import model.Character;
import model.Rescale;

public class Graph extends View {
    String templateName;

    String title;
    String date;
    String entrants;
    String location;

    ArrayList<Player> players;

    Data characterData;
    int IMAGE_HEIGHT = 500;

    BitmapDrawable drawable;

    public Graph(Context context, String template, String title, String date, String entrants, String location, ArrayList<Player> players) {

        super(context);
        this.templateName = template;
        this.title = title;
        this.date = date;
        this.entrants = entrants;
        this.location = location;
        this.players = players;
        this.characterData = new Data();

        this.createDrawable();
    }

    public void createDrawable(){
        // Bitmap création
        Bitmap bitmap = Bitmap.createBitmap(1600, 2331, Bitmap.Config.ARGB_8888);
        Canvas c = new Canvas(bitmap);

        // Template background
        int id = getContext().getResources().getIdentifier("template_" + this.templateName, "drawable", getContext().getPackageName());
        Drawable background = getContext().getResources().getDrawable(id);
        Rect bounds = c.getClipBounds();
        background.setBounds(bounds);
        background.draw(c);

        // Title
        Paint titleText = new TextPaint();
        titleText.setAntiAlias(true);
        titleText.setTextAlign(Paint.Align.CENTER);
        titleText.setTypeface(ResourcesCompat.getFont(getContext(), R.font.bebas_neue));
        titleText.setTextSize(70);

        if (this.templateName.matches("man")) {
            titleText.setColor(ContextCompat.getColor(getContext(), R.color.title_man));
        } else if (this.templateName.matches("mdtk")) {
            titleText.setColor(ContextCompat.getColor(getContext(), R.color.title_mdtk));
        } else if (this.templateName.matches("sun")) {
            titleText.setColor(ContextCompat.getColor(getContext(), R.color.title_sun));
        } else if (this.templateName.matches("cash")) {
            titleText.setColor(ContextCompat.getColor(getContext(), R.color.title_cash));
        }

        c.drawText(this.title,  800, 560, titleText);

        // Texts
        Paint text = new TextPaint();
        text.setAntiAlias(true);
        text.setTextAlign(Paint.Align.LEFT);
        text.setTypeface(ResourcesCompat.getFont(getContext(), R.font.bebas_neue));
        text.setColor(Color.WHITE);
        text.setTextSize(64);

        c.drawText(this.entrants,  125, 448, text);
        c.drawText(this.location,  1250, 448, text);
        c.drawText(this.date,  754, 2312, text);

        // Player text
        Paint playerText = new TextPaint();
        playerText.setAntiAlias(true);
        playerText.setTextAlign(Paint.Align.LEFT);
        playerText.setTypeface(ResourcesCompat.getFont(getContext(), R.font.bebas_neue));
        playerText.setColor(Color.WHITE);
        playerText.setTextSize(150);

        for (int i=0; i<8; i++) {
            int[] y = {811, 1025, 1234, 1441, 1654, 1840, 2055, 2241}; // The gap between layout aren't equals
            Player player = this.players.get(i);

            // Characters
            // Remove null character
            ArrayList<Character> toRemove = new ArrayList<>();
            player.characters.remove(null);
            for (Character character : player.characters) {
                int characterId = getContext().getResources().getIdentifier(character.getPath(templateName), "drawable", getContext().getPackageName());
                if (characterId == 0) {
                    toRemove.add(character);
                }
            }
            player.characters.removeAll(toRemove);

            int[] posX = new int[]{1200};;
            int posY = y[i];

            ArrayList<Integer> widthArray = new ArrayList<>();
            int totalWidth = 0;
            for (Character character : player.characters) {
                int characterId = getContext().getResources().getIdentifier(character.getPath(templateName), "drawable", getContext().getPackageName());
                if (characterId == 0) break;
                Bitmap characterBitmap = BitmapFactory.decodeResource(getContext().getResources(), characterId);

                // Rescale and crop the image
                int initialWidth = characterBitmap.getWidth();
                int initialHeight = characterBitmap.getHeight();

                Rescale rescale = characterData.getRescale(character.getPath(templateName).split("_alt")[0]);

                int rescaleWidth = rescale.width * initialHeight / IMAGE_HEIGHT; // characterBitmap.getHeight != original image height

                characterBitmap = Bitmap.createBitmap(characterBitmap,
                        0,
                        (i == 0 ? 0 : rescale.cropTop), // Character not rescaled for the first player
                        initialWidth,
                        (initialHeight - rescale.cropBottom - (i == 0 ? 0 : rescale.cropTop)));

                int height = 169 + (i == 0 ? rescale.cropTop * 169 / characterBitmap.getHeight() : 0 );
                int reduceGap = (player.characters.size() == 3 ? 50 : 0); // reduce gap by 50 if 3 characters
                int width = (rescaleWidth * height) / characterBitmap.getHeight() - reduceGap;
                widthArray.add(width);
                totalWidth += width;
            }

            switch(player.characters.size()) {

                case 1:
                    posX = new int[]{1200 - (totalWidth / 2)};
                    break;

                case 2:
                    posX = new int[]{1200 - (totalWidth / 2), 1200 - (totalWidth / 2) + widthArray.get(0)};
                    break;

                case 3:
                    posX = new int[]{1200 - (totalWidth / 2),
                            1200 - (totalWidth / 2) + widthArray.get(0),
                            1200 - (totalWidth / 2) + widthArray.get(0) + widthArray.get(1)};
                    break;
            }

            for (int index = player.characters.size()-1; index>=0; index--) { // draw the character in reverse order for the layer level
                Character character = player.characters.get(index);
                // Get drawable
                int characterId = getContext().getResources().getIdentifier(character.getPath(templateName), "drawable", getContext().getPackageName());
                Bitmap characterBitmap = BitmapFactory.decodeResource(getContext().getResources(), characterId);

                // Rescale and crop the image
                int initialWidth = characterBitmap.getWidth();
                int initialHeight = characterBitmap.getHeight();

                Rescale rescale = characterData.getRescale(character.getPath(templateName).split("_alt")[0]);

                characterBitmap = Bitmap.createBitmap(characterBitmap,
                        0,
                        (i == 0 ? 0 : rescale.cropTop), // Character not rescaled for the first player
                        initialWidth,
                        (initialHeight - rescale.cropBottom - (i == 0 ? 0 : rescale.cropTop)));

                int height = 169 + (i == 0 ? rescale.cropTop * 169 / characterBitmap.getHeight() : 0 );
                int width = (characterBitmap.getWidth() * height) / characterBitmap.getHeight();
                int horizontalOffset = rescale.horizontalOffset * width / characterBitmap.getWidth();
                Drawable characterDraw = new BitmapDrawable(getResources(), characterBitmap);

                // Set bounds
                characterDraw.setBounds(posX[index] - horizontalOffset,
                        posY - height,
                        posX[index] + width - horizontalOffset,
                        posY);

                characterDraw.draw(c);
            }

            // Team
            if (player.team != null) {
                int teamId = getContext().getResources().getIdentifier("team_" + player.team, "drawable", getContext().getPackageName());
                if (teamId != 0) {
                    Bitmap teamBitmap = BitmapFactory.decodeResource(getContext().getResources(), teamId);

                    // Rescale
                    int height = 120;
                    int width = (teamBitmap.getWidth() * height) / teamBitmap.getHeight();
                    Drawable team = new BitmapDrawable(getResources(), teamBitmap);

                    team.setBounds(340 - (width / 2),
                            posY - height - 22,
                            340 + (width / 2),
                            posY - 22);
                    team.draw(c);
                }
            }

            // Pseudo
            c.drawText(player.pseudo, 425, posY-30, playerText);

        }

        // Set canvas bitmap to background
        drawable = new BitmapDrawable(getResources(), bitmap);
    }

    @Override
    protected void onDraw(Canvas c) {
        super.onDraw(c);
        this.setBackground(drawable);
    }
}

File: app/src/main/java/com/ovdr/graphmaker/model/Graph.kt
package com.ovdr.graphmaker.model

import android.content.Context
import android.graphics.*
import android.graphics.drawable.BitmapDrawable
import android.text.TextPaint
import android.view.View
import androidx.core.content.ContextCompat
import androidx.core.content.res.ResourcesCompat
import com.ovdr.graphmaker.R

class Graph(
    context: Context,
    private val title: String,
    private val date: String,
    private val entrants: String,
    private val location: String,
    private val players: ArrayList<Player>
    ): View(context) {
    private lateinit var drawable: BitmapDrawable

    init {
        createDrawable()
    }

    private fun createDrawable() {
        // Bitmap creation
        val bitmap = Bitmap.createBitmap(1600, 2263, Bitmap.Config.ARGB_8888)
        val c = Canvas(bitmap)

        // Template background
        val background = ContextCompat.getDrawable(context, R.drawable.template_man_background)!!
        val bounds = c.clipBounds
        background.bounds = bounds
        background.draw(c)

        // Players
        val playerText: Paint = TextPaint()
        playerText.isAntiAlias = true
        playerText.typeface = ResourcesCompat.getFont(context, R.font.montserrat_bold)
        playerText.textAlign = Paint.Align.LEFT
        playerText.textSize = 100f

        // Character scale map
        val characterMap = resources.getStringArray(R.array.character_array).associateBy({ it.split("|")[0]}, {it.split("|")[1]})

        for ((playerIndex, player) in players.withIndex()) {
            // Characters
            val characters = player.characters.filter { character -> character !== ""}
            var offset = 0
            for (character in characters.reversed()) {
                // Get drawable
                val characterId = context.resources.getIdentifier(
                    "character_$character",
                    "drawable",
                    context.packageName
                )

                // Create bitmap
                var characterBitmap = BitmapFactory.decodeResource(context.resources, characterId)
                val width = characterBitmap.width
                val height = characterBitmap.height

                // Crop
                val data = characterMap[character]!!
                val x = data.substringAfter("(").substringBefore(";").toInt() * height / 500
                val y = data.substringAfter(";").substringBefore(")").toInt() * height / 500
                val h = data.substringAfter(":").substringBefore("-").toInt() * height / 500
                val w = data.substringAfter("-").toInt() * height / 500

                characterBitmap = Bitmap.createBitmap(characterBitmap,
                        0,
                        maxOf(0, (y-(h/2))),
                        width,
                        minOf(height-(y-(h/2)), h)
                    )

                // Scale and draw
                val characterDrawable = BitmapDrawable(resources, characterBitmap)
                characterDrawable.setBounds(
                    (1250+50*characters.size) - (x+w/2)*153/h - offset,
                    578+playerIndex*(35+153),
                    (1250+50*characters.size) - offset + (width-x-w/2)*153/h,
                    578+playerIndex*(35+153) + 153,
                )
                offset += w*153/h +  (60/characters.size)
                characterDrawable.draw(c)
            }

            // Shadow
            playerText.color = ContextCompat.getColor(context, R.color.green_400)
            c.drawText(player.pseudo, 405f, 690f+playerIndex*(35f+153f), playerText)
            // Pseudo
            playerText.color = Color.WHITE
            c.drawText(player.pseudo, 400f, 685f+playerIndex*(35f+153f), playerText)
        }

        // Template foreground
        val foreground = ContextCompat.getDrawable(context, R.drawable.template_man_foreground)!!
        foreground.bounds = bounds
        foreground.draw(c)

        // Title
        val titleText: Paint = TextPaint()
        titleText.isAntiAlias = true
        titleText.typeface = ResourcesCompat.getFont(context, R.font.montserrat_bold)
        titleText.color = ContextCompat.getColor(context, R.color.green_400)
        titleText.textAlign = Paint.Align.CENTER
        titleText.textSize = 70f
        c.drawText(title, 800f, 520f, titleText)

        // Texts
        val text: Paint = TextPaint()
        text.isAntiAlias = true
        text.typeface = ResourcesCompat.getFont(context, R.font.montserrat_bold)
        text.color = Color.WHITE
        text.textAlign = Paint.Align.LEFT
        text.textSize = 64f
        c.drawText(location, 66f, 200f, text)
        c.drawText(entrants, 66f, 275f, text)
        text.textAlign = Paint.Align.RIGHT
        c.drawText(date, 1535f, 275f, text)

        // Set canvas bitmap to background
        drawable = BitmapDrawable(resources, bitmap)
    }

    override fun draw(c: Canvas?) {
        super.draw(c)
        this.background = drawable
    }
}